ledger/acctupdates.go

```diff
@@ -896,28 +926,140 @@ func (au *accountUpdates) initializeCaches(lastBalancesRound, lastestBlockRound,
+			case blocksStream <- blk:
+			case <-blockEvalFailed:
+				return
+			}
+		}
+	}()
+
+	lastFlushedRound := lastBalancesRound
+	const accountsCacheLoadingMessageInterval = 5 * time.Second
+	lastProgressMessage := time.Now().Add(-accountsCacheLoadingMessageInterval / 2)
+
+	// rollbackSynchronousMode ensures that we switch to "fast writing mode" when we start flushing out rounds to disk, and that
+	// we exit this mode when we're done.
+	rollbackSynchronousMode := false
+	defer func() {
+		if rollbackSynchronousMode {
+			// restore default synchronous mode
+			au.dbs.Wdb.SetSynchronousMode(context.Background(), au.synchronousMode, au.synchronousMode >= db.SynchronousModeFull)
 		}
+	}()
 
+	for blk := range blocksStream {
 		delta, err = au.ledger.trackerEvalVerified(blk, &accLedgerEval)
 		if err != nil {
+			close(blockEvalFailed)
 			return
 		}
 
 		au.newBlockImpl(blk, delta)
-		lastBalancesRound = next
 
-		if next == basics.Round(writingCatchpointRound) {
+		if blk.Round() == basics.Round(writingCatchpointRound) {
 			catchpointBlockDigest = blk.Digest()
 		}
 
+		// flush to disk if any of the following applies:
+		// 1. if we have loaded up more than initializeCachesRoundFlushInterval rounds since the last time we flushed the data to disk
+		// 2. if we completed the loading and we loaded up more than 320 rounds.
+		if blk.Round()-lastFlushedRound > initializeCachesRoundFlushInterval ||
+			(lastestBlockRound == blk.Round() && lastBalancesRound+basics.Round(blk.ConsensusProtocol().MaxBalLookback) < lastestBlockRound) {
+			// adjust the last flush time, so that we would not hold off the flushing due to "working too fast"
+			au.lastFlushTime = time.Now().Add(-balancesFlushInterval)
+
+			if !rollbackSynchronousMode {
+				// switch to rebuild synchronous mode to improve performance
+				au.dbs.Wdb.SetSynchronousMode(context.Background(), au.accountsRebuildSynchronousMode, au.accountsRebuildSynchronousMode >= db.SynchronousModeFull)
+
+				// flip the switch to rollback the synchronous mode once we're done.
+				rollbackSynchronousMode = true
+			}
+
+			// The unlocking/relocking here isn't very elegant, but it does get the work done :
+			// this method is called on either startup or when fast catchup is complete. In the former usecase, the
+			// locking here is not really needed since the system is only starting up, and there are no other
+			// consumers for the accounts update. On the latter usecase, the function would always have exactly 320 rounds,
+			// and therefore this wouldn't be an issue.
+			// However, to make sure we're not missing any other future codepath, unlocking here and re-locking later on is a pretty
+			// safe bet.
+			au.accountsMu.Unlock()
+
+			// flush the account data
+			au.committedUpTo(blk.Round())
+
+			// wait for the writing to complete.
+			au.waitAccountsWriting()
+
+			// The au.dbRound after writing should be ~320 behind the block round.
+			roundsBehind := blk.Round() - au.dbRound
```

Comment:
Is it true for the case when accounts behind blocks for 1M rounds? If not how would not the condition below about the process?

Reply:
on "normal" use case, after writing the accounts for a given block, the au.dbRound would be at most 320 rounds behind that block. The goal of this test is to see that there hasn't been any errors generated during the `commitRound` execution. If there were any, there is no point in keep going.
As the node below suggest, a catchpoint can tweak that, so we need to take that into account as well. This "issue" comes because we're not calling `committedUpTo` for every round, but rather in 1000 rounds "jumps".